File: 0001-0050/0006/solution.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        rows = [""] * numRows
        i, row, down = 0, 0, False
        for i in range(len(s)):
            rows[row] += s[i]
            if row == 0 or row == numRows - 1:
                down = not down
            row += 1 if down else -1
        ans = ""
        for row in rows:
            ans += row
        return ans


class Example:
    def __init__(self, s, numRows, ans):
        self.s = s
        self.numRows = numRows
        self.ans = ans


def main():
    examples = [
        Example("PAYPALISHIRING", 3, "PAHNAPLSIIGYIR"),
        Example("PAYPALISHIRING", 4, "PINALSIGYAHRPI"),
        Example("A", 1, "A")
    ]

    solve = Solution()
    for i, example in enumerate(examples):
        ans = solve.convert(example.s, example.numRows)
        if ans == example.ans:
            print(f"PASS: CASE {i}")
        else:
            print(f"FAIL: CASE {i}")


if __name__ == "__main__":
    main()

File: 0001-0050/0006/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1) return s;
        vector<string> rows(numRows, "");
        bool down = false;
        for (int i = 0, row = 0; i < s.length(); i++) {
            rows[row] += s[i];
            if (row == 0 || row == numRows - 1) down = !down;
            row += down ? 1 : -1;
        }
        string ans = "";
        for (int i = 0; i < numRows; i++) ans += rows[i];
        return ans;
    }
};

struct Example {
    string s;
    int numRows;
    string ans;
};

int main() {
    vector<Example> examples = {
        {"PAYPALISHIRING", 3, "PAHNAPLSIIGYIR"},
        {"PAYPALISHIRING", 4, "PINALSIGYAHRPI"},
        {"A", 1, "A"},
    };
    
    Solution *solve = new Solution();
    for (int i = 0; i < examples.size(); i++) {
        string ans = solve->convert(examples[i].s, examples[i].numRows);
        if (ans == examples[i].ans) {
            cout << "PASS: CASE " << i << endl;
        } else {
            cout << "FAIL: CASE " << i << endl;
        }
    }
    
    return 0;
}